### DISCUSSION TITLE
Will json.dump() get the indent argument?

### DISCUSSION THREAD

**Discussion by liudr** [micropython/micropython#16296] [Core Development] [answered]
I am wondering if adding "indent" argument to json.dump() may be in any plan. My use case is to write json strings to the dev board so it can be later opened say in Thonny and easily modified by hand. Currently there is no indent option so everything is turned into a single line of string, impossible to read.

If there is no plan to add this feature, is there any easy way to obtain this feature by other means? Thank you!

**Comment by Josverl** [comment, marked as answer]
There have been a few PRs to try to reduce the size of the generated json by omitting spaces  [#4294](https://github.com/micropython/micropython/pull/4294#issuecomment-842849492) , #7337 
adding `ident=4` would be inline with CPython , but I thing the code for that is non-trivial and will increase the firmware size.

Eve after that - there is a pitfall waiting to happen if you try to 
- write a file on the device 
- read , modify and save that from the host [ Assuming MSC ] 
- try to read it from your device 

This is **very likely** to cause filesystem corruption , see #16012 for one of the recent conversations on this topic.

So my answer is : Quite unlikely 

If you know something about the json structure, you may be able to save memory , and add indentation at the same time.
I have used that for a few years already in [MicroPython-stubber - createstubs.py](https://github.com/Josverl/micropython-stubber/blob/a3c2486d11468ef87d350f15b7e6403371824350/src/stubber/board/createstubs.py#L428-L464).

The gist: 
```py
# write 'header'
            # write json by node to reduce memory requirements
            with open(self._json_name, "w") as f:
                f.write("{")
                f.write(dumps({"firmware": self.info})[1:-1])
                f.write(",\n")
                f.write(dumps({"stubber": {"version": __version__}, "stubtype": "firmware"})[1:-1])
                f.write(",\n")
                f.write('"modules" :[\n')

# Write data array 
            with open(self._json_name, "a") as f:
                if not self._json_first:
                    f.write(",\n")
                else:
                    self._json_first = False
                # Add indentation to the line below
                indent = 4
                line = '{{{}"module": "{}", "file": "{}"}}'.format(" "*indent, module_name, stub_file.replace("\\", "/"))
                f.write(line)

# Write 'trailer'
        with open(self._json_name, "a") as f:
            f.write("\n]}")

```

**Reply by liudr** [reply]
Thanks for your immediate response! I was expecting NO as well, for the exact same reason to minimize firmware size. To clarify, I am using MicroPython and not CircuitPython so there is no MSC while MicroPython is running. My use case can be more specific to ESP32 (no S variants) so file corruption shouldn't be a concern here.

Since I'm using ESP32-WROVER, I have quite a bit of resources (PSRAM and FLASH). Is there a way to somehow install a CPython json module at all? My json text is pretty complex, with a dict that includes key:value pairs that each includes multiple levels of dicts and lists. I don't know if the stubber functions are meant for that. Maybe one data array with some header key:value pairs but not what I am doing.

**Reply by andrewleech** [reply]
It might be straightforward to port this one: https://pypi.org/project/simplejson/

**Reply by liudr** [reply]
Thanks @andrewleech I took a stab at that module but wasn't able to get it to work. It imports __future__ probably because it says it wants to stay backward compatible with Python 2.7 and also uses at least one module we don't have, decimal. I'm sure if I look further there's more modules that can't be resolved. For now I have these ways:

1. Ignore the fact that json.dump() outputs horrible 1-liner for a once easily human-readable json file until someone told me it's not good enough and has to be pretty.
2. Do a string operation to read in the entire json file and assume that the key:value pair I want to update stays in a single line so I can replace the value without changing the format. I don't like this. Too specific, no generalization to any future updates on other values.
3. Well, if anyone isn't happy with the 1-line format, tell them to beautify it with an online json beautifier and save it back to the board. Since they are complaining, they must have console access to the board. ;)

**Reply by stinos** [reply]
> beautify it with an online json beautifier 

All modern text editors support pretty printing JSON, so maybe Thonny can also get a plugin for that?

**Reply by jgriesfeller** [reply]
you might want to have a look at `jq` (https://jqlang.github.io/jq/) for pretty printing.

**Reply by Josverl** [reply]
> I am using MicroPython and not CircuitPython so there is no MSC while MicroPython is running.


MSC is an option with micropython as well, but without the additional safeties that CP has added to avoid dual ownership of the vfs. So not for the unaware.

**Reply by liudr** [reply]
> > I am using MicroPython and not CircuitPython so there is no MSC while MicroPython is running.
> 
> MSC is an option with micropython as well, but without the additional safeties that CP has added to avoid dual ownership of the vfs. So not for the unaware.

Noted. Thanks. As I stated I'm primarily concerned with ESP32 so there is only USB CDC from CP2102/N. The board either is running a script that has sole control of the vfs or I interrupt the script to enter REPL where I use Thonny to modify the vfs. I don't think there is any danger of dual ownership, is there?
